### PR TITLE
[20629] GUIDLess Discovery Server

### PR DESCRIPTION
**Pull request by cferreiragonz**
## Description

This PR drastically changes Discovery Server behavior to be able to work without a fixed GUID. 

- Servers and Clients now connect between each other by exclusively specifying a locator, the GUID is not needed anymore. The remote server list has been replaced with LocatorList.
- Hence, the server_id parameter used in the CLI and the positional arguments in the environment variable are deprecated. The -i parameter in the CLI has been kept in order to guarantee backward compatibility, but it is not mandatory any longer. 
- Servers now use a mesh topology, meaning that every server will directly connect with all of the available servers. 
- In order to accomplish last point, Servers now send the Data(p) of connected servers to new servers connections and send the Data(p) of the new server to all connected servers. To connect a new server to the mesh, it is only needed to specify the locator of an already connected server.
- Servers only redirect discovery information of their direct clients, and will stop redirecting discovery information of the clients of other servers.
- Remote Discovery servers connection list can now be updated at runtime without any restrictions. The previous list does not need to be a subset of the new one. 
- Servers now use the lease announcements routine to make connections with the remote servers specified in the remote Discovery servers list. Initial announcements will be called again every time a known server is dropped or the remote server list is updated.
- Clients now accept connections of any server, not only the ones specified in their remote Discovery servers list. Although in normal behavior a client will only receive discovery information of the servers pinged by itself, if it receives a Data(p) of an other server (sending it manually) it will try to make a connection.
- ServerAttributes is now private, as it is no longer needed to configure Clients or Servers

- This PR is on top of https://github.com/eProsima/Fast-DDS/pull/4740 and must be merged after it
- It also affects the python API: https://github.com/eProsima/Fast-DDS-python/pull/147

## Contributor Checklist

- [X] Commit messages follow the project guidelines. 
- [X] The code follows the style guidelines of this project. 
- [X] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally: https://github.com/eProsima/Discovery-Server/pull/76 
- [X] Any new/modified methods have been properly documented using Doxygen. 
- [X] Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- ❌ Changes are ABI compatible. 
- ❌ Changes are API compatible. (RemoteServers are now specified by a LocatorList. Also, Fast DDS tool CLI and ROS_DISCOVERY_SERVER env. var. behavior have been modified)
- [x] New feature has been added to the `versions.md` file (if applicable).
- [X] New feature has been documented/Current behavior is correctly described in the documentation. Please uncomment following line with the corresponding PR to the documentation project: -->
    - Related documentation PR: https://github.com/eProsima/Fast-DDS-docs/pull/751
- _N/A_ Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/rtps/builtin/discovery/participant/PDPServerListener.hpp [hunk 1]
```diff
@@ -65,12 +65,14 @@ class PDPServerListener : public fastrtps::rtps::PDPListener
      * Checks discovery conditions on a discovery server entity.
      * Essentially, it checks for incoming PIDS of remote proxy datas.
      * @param participant_data Remote participant data to check.
+     * @param participant_type_str Type of the remote participant.
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDPServerListener.hpp:68]:
```suggestion
     * @param [in]  participant_data Remote participant data to check.
     * @param [out] participant_type_str Type of the remote participant.
```

File: src/cpp/rtps/builtin/discovery/participant/PDPServerListener.hpp [hunk 2]
```diff
@@ -65,12 +65,14 @@ class PDPServerListener : public fastrtps::rtps::PDPListener
...
      * @return A pair of booleans.
      * The first one indicates if the remote participant data is valid.
      * The second one indicates if the remote participant data is a client.
      */
     std::pair<bool, bool> check_server_discovery_conditions(
-            fastrtps::rtps::ParticipantProxyData& participant_data);
+            fastrtps::rtps::ParticipantProxyData& participant_data,
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDPServerListener.hpp:74]:
```suggestion
            const fastrtps::rtps::ParticipantProxyData& participant_data,
```

File: src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp [hunk 3]
```diff
@@ -188,6 +174,7 @@ void PDPServer::pre_enable_actions()
     if (durability_ == TRANSIENT)
     {
         // This vector is empty till backup queue is implemented
+        // TODO Carlos (OPT-IMPROVE): delete this?
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp:177]: ??

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: test/mock/rtps/RTPSParticipantAttributes/fastdds/rtps/attributes/RTPSParticipantAttributes.h [hunk 4]
```diff
@@ -41,6 +40,8 @@
 #include <fastdds/rtps/transport/UDPv4TransportDescriptor.h>
 #include <fastdds/fastdds_dll.hpp>
 
+#include <rtps/attributes/ServerAttributes.hpp>
+
```
Comment (MiguelCompany) [inline test/mock/rtps/RTPSParticipantAttributes/fastdds/rtps/attributes/RTPSParticipantAttributes.h:44]:
```suggestion
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/RTPSDomain.cpp [hunk 5]
```diff
@@ -36,6 +36,7 @@
 #include <fastdds/utils/IPLocator.h>
 #include <fastdds/utils/md5.h>
 
+#include <rtps/attributes/ServerAttributes.hpp>
```
[the remark this answers is not in the document]
Comment (cferreiragonz) [reply src/cpp/rtps/RTPSDomain.cpp:39]: We need it to use `load_environment_server_info`, don't we?

File: src/cpp/rtps/attributes/ServerAttributes.cpp [hunk 6]
```diff
@@ -0,0 +1,445 @@
+// Copyright 2022 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
Comment (MiguelCompany) [inline src/cpp/rtps/attributes/ServerAttributes.cpp:1]:
```suggestion
// Copyright 2024 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```

File: src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp [hunk 7]
```diff
@@ -933,412 +933,6 @@ void PDPClient::match_pdp_reader_nts_(
...
-    {
-        EPROSIMA_LOG_ERROR(SERVER_CLIENT_DISCOVERY, e.what());
-        servers_list.clear();
-        return false;
-    }
-
-    return true;
-}
-
 GUID_t RemoteServerAttributes::GetParticipant() const
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp:936]: Implementation of `RemoteServerAttributes` methods shall be moved to `ServerAttributes.cpp`

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — DISMISSED** [review]
LGTM with green CI

**Review by MiguelCompany — APPROVED** [review]

**Comment by cferreiragonz** [comment]
Green CI. Updating `versions.md` and RTM

**Pull request merged**